ANA-434: LifeCycle Management Of FxFwds

jz-finbourne opened the pull request:
# Pull Request Checklist

- [x] Read the [contributing guidelines](../blob/master/docs/CONTRIBUTING.md)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

Adds a test demonstrating the lifecycle management of FxForwards. That is, the sum PV of the forward and its cashflows is constant over time (assuming constant FX rates and no discounting)

rizwansaeed commented:
Its not clear to me what we're trying to show here to the reader.  It looks like a perfectly good "test" but that should probably live in the main LUSID project. The idea behind the sdk **examples** is that they demonstrate features with a focus on showing a potential user what code to write, that than terseness of code or proving LUSID capabilities.

From what I can infer we are trying to show 
* that a forward expires into cash
* we can get the cashflows
* we book those as transactions

If that is the case we should probably show
* the portfolio contents before expiry
* the cashflows we are going to book
* the portfolio contents after booking

jz-finbourne commented:
> Its not clear to me what we're trying to show here to the reader. It looks like a perfectly good "test" but that should probably live in the main LUSID project. The idea behind the sdk **examples** is that they demonstrate features with a focus on showing a potential user what code to write, rather than terseness of code or proving LUSID capabilities.
> 
> From what I can infer we are trying to show
> 
> * that a forward expires into cash
> * we can get the cashflows
> * we book those as transactions
> 
> If that is the case we should probably show
> 
> * the portfolio contents before expiry
> * the cashflows we are going to book
> * the portfolio contents after booking



> Its not clear to me what we're trying to show here to the reader. It looks like a perfectly good "test" but that should probably live in the main LUSID project. The idea behind the sdk **examples** is that they demonstrate features with a focus on showing a potential user what code to write, rather than terseness of code or proving LUSID capabilities.
> 
> From what I can infer we are trying to show
> 
> * that a forward expires into cash
> * we can get the cashflows
> * we book those as transactions
> 
> If that is the case we should probably show
> 
> * the portfolio contents before expiry
> * the cashflows we are going to book
> * the portfolio contents after booking

Hi Riz,

Thank you for your comments. You are correct above in what we are trying to demonstrate.  Perhaps the code is/was not very clear, in my opinion it was doing :

- the portfolio contents before expiry
- the cashflows we are going to book
- the portfolio contents after booking

I was checking portfolio contents by calling GetValuation() - would you suggest another / better way? [GetValuation a couple days before, day of, and 2 days after expiry and the contents/pv are checked before/during/after expiry is how I am checking the portfolio content and its pv].
 
In terms of the cashflows we are booking, I did not want to check these numerically. I have added assertions on its currency. The only implicit numeric assertion part is checking the PV is conserved along time.

I have added some extra comments to make clearer to hopefully make clearer what is going on.

jz-finbourne commented on sdk/Lusid.Sdk.Tests/tutorials/Ibor/PortfolioCashFlows.cs:
```diff
@@ -261,6 +279,125 @@ public void ExampleUpsertablePortfolioCashFlowsForFxForwards(bool isNdf)
+                .Select(pv => Math.Round(pv, 12))
+                .Distinct()
+                .ToList();
+            Assert.That(uniquePvsAcrossDates.Count, Is.EqualTo(1));
```
you mean something like:
```
var firstPrice = x.First();
foreach(price in Prices)
{
  Assert.that(price, Is.EqualTo(firstPrice))
}
```
to explicitly show the price is same throughout (than abusing `.Distinct()` trick)?
